main/mqtt.c

```diff
@@ -106,8 +133,43 @@ mqtt_publish_connect(void)
     const int32_t mqtt_qos         = 1;
     const int32_t mqtt_flag_retain = 1;
 
-    const mqtt_message_id_t message_id
-        = esp_mqtt_client_publish(gp_mqtt_client, topic.buf, p_message, strlen(p_message), mqtt_qos, mqtt_flag_retain);
+    esp_mqtt_client_handle_t p_mqtt_client = mqtt_mutex_lock();
+    const mqtt_message_id_t  message_id    = esp_mqtt_client_publish(
+        p_mqtt_client,
+        topic.buf,
+        p_message,
+        (int)strlen(p_message),
+        mqtt_qos,
+        mqtt_flag_retain);
+    mqtt_mutex_unlock(&p_mqtt_client);
+    if (-1 == message_id)
+    {
+        LOG_ERR("esp_mqtt_client_publish failed");
+    }
+    else
+    {
+        LOG_INFO("esp_mqtt_client_publish: message_id=%d", message_id);
+    }
+}
+
+static void
+mqtt_publish_state_offline(esp_mqtt_client_handle_t p_mqtt_client)
+{
+    char *p_message = "{\"state\": \"offline\"}";
+
+    mqtt_topic_buf_t topic;
+    mqtt_create_full_topic(&topic, g_gateway_config.mqtt.mqtt_prefix, "gw_status");
+    LOG_INFO("esp_mqtt_client_publish: topic:'%s', message:'%s'", topic.buf, p_message);
+    const int32_t mqtt_qos         = 0;
+    const int32_t mqtt_flag_retain = 0;
```

Comment:
The status message for online/offline could have retain_flag = 1, it would make sense to me that new client connecting would get retained gateway statuses.

Reply:
status message "online" has retain flag, but I'm not sure if it's needed for "offline" message because it is sent only when new configuration is sent to gw. 
When the gateway is reconfigured, MQTT client is stopped before applying new configuration and gw_status="offline" is sent at this moment.

Reply:
It seems to me that there's 3 possible states for MQTT:

1) Gateway is online and configured to use MQTT
2) Gateway is working and connected to Internet, but configured to not use MQTT / connect to different broker
3) Gateway is offline due to power / connectivity loss, LWT message state. 

It would make sense to have all these states in one topic and retained, there should not be any significant cost for the broker to retain one topic per gateway.

Reply:
ok